### DISCUSSION TITLE
How to DI generic objects from a container?

### DISCUSSION THREAD

**Discussion by exteon** [symfony/symfony#42417] [Q&A]
Hi all, I have the following problem to fix, hopefully with symfony:
I have a bunch of domain model classes (each of which can be instanced to any infinite number of instances).
Some of these classes rely on services to run, that should be provided by a container (i.e. `Mailer`)
At this point, all instances hold a reference to a container to demand the services. I'm trying to eliminate this backlink.
In order to do that, I'm looking for something akin' to Laravel's `App::make()`, namely let's say I have:

```php
class MailMan {
    public function __construct(Mailer $mailer){
    }
}
```

I'm looking for a container builder that can build me a `MailMan` with the `Mailer` dependency resolved.
I looked at https://github.com/symfony/dependency-injection/blob/a665946279f566d94ed5eb98999cfa65c6fa5a78/ContainerBuilder.php#L545 (`doGet()`) and it only seems to resolve dependencies between services and instantiate a singleton service.
From further reading the source code, it appears what I need is somewhere down the line of (I haven't tried it yet):
```php
public function abstractBuilder(string $class) {
    $definition = new Definition($class);
    $args = $this->containerBuilder->resolveServices($definition->getArguments());
   return new $class($args);
}
```
Is there another standard way of achieving this with Symfony?
Thanks!

**Comment by ngrie** [comment]
I don't think I completely understood what you are trying to achieve, but this sounds wrong to me:

> I have a bunch of domain model classes (each of which can be instanced to any infinite number of instances).
Some of these classes rely on services to run

A domain object should not depend on a service. If your logic depends on a service, it should be located in a service itself and use autowiring to resolve its dependencies.

**Reply by exteon** [reply]
https://github.com/symfony/symfony/discussions/42417#discussioncomment-1142940 I don't know how to write a reply to both your answers :)

**Comment by wouterj** [comment]
The concept of dependency injection (container) in Laravel and Symfony is quite different (and mostly not even comparable). In Laravel, the container acts mostly as a "factory" to create classes using `App::make()`. In Symfony, this is not supported and also unwanted.

Like @ngrie said, in Symfony something must be a service if it relies on services (and thus must also be only retrieved using dependency injection). This forces you to write code that doesn't depend on the container (i.e. no class in your domain should be aware that there is something like a container).

Btw, I would not recommend to use the `abstractBuilder()` code snippet you showed in production. The container builder is not meant to be used at runtime and is most likely to cause a bad performance. In Symfony, as all services are known from their configuration, the container builder is dumped into an performance optimized PHP file in the `cache/` directory. Using the container builder in production will by-pass this.

**Reply by exteon** [reply]
https://github.com/symfony/symfony/discussions/42417#discussioncomment-1142940 I don't know how to write a reply to both your answers :)

**Comment by exteon** [comment]
I think I got confused and confusing by trying to oversimplify things; let me elaborate a bit:
- I have a large number of domain classes, let's call them `Dom1` ... `Dom200`
- They are all built (at the current moment) by an abstract factory/manager: 
  ```php
    /** @var ObjectManager $objectManager **/
    $instance = $objectManager->create(Dom1::class);
  ```
Each of these domain classes have various dependencies, some of them cross-cutting (let's have, for example `LanguageContext`, `Logger`, `CachingRssReader` etc).

The way they are wired at the moment is that each of the domain instances holds a reference to the `ObjectManager` to require their dependencies, so `ObjectManager` acts more or less like a "service". Except for me, semantically, there's no difference between a "service" and a "container": a "service" is just a container within a container; and I'll explain why:

By requiring domain instances to hold a reference to a "locally global" object, it creates a mess of testability and portability: by the very same principle that a domain object should be oblivious there is a "container", it should be oblivious that there is a "service".

Now, I have two was to get out of this mess:

1) Create specialized builders
  If I replace `ObjectManager` with `Dom1Factory` ... `Dom200Factory`, each factory will forward the needed dependencies from the "service" (or whatever other form of container, it doesn't really matter), to the domain object's constructors.
  However, I am unwilling to do so, as the effort of adding and maintaining the 200 specialized builders is unjustified

2) Have an abstract builder within the "service" parse the domain object's constructor and autowire it to the "service"'s state; **I do not need this to wire to an different service as it may seem my initial question implied**, I just need it to wire to some form of container (be it a "service" or a "container" or an "App::", as long as I don't have to create 200 specialized builders for each entity's signature)

I just pointed to the hack I saw because it looked to me like the functionality is already there, it just doesn't seem to be granular enough to handle anything smaller than a "service". So basically, to reset the talk, I just need a form of container with a form of abstract builder that can autowire generic objects, without the need for me to call them "services" or anything other than "objects".

Ps. also, sorry for the mess, I don't use this Github chat much, how do I post a reply to both of you that is not an "answer"? It seems I just wrote this as an answer to my question :)

**Reply by ro0NL** [reply]
while im generally skeptical about (domain) models, depending on services (usually it's in reverse; services provide/take models)

there's a way to leverage Symfony DI as a general purpose factory:

```php
/** @var Some $instance */
$instance = $container->get(Some::class);
```


This allows to mass autowire say `Dom001 ~ Dom200` using service disocvery: https://symfony.com/doc/current/service_container.html#importing-many-services-at-once-with-resource

to make it truly a factory, you want a new instance each time: https://symfony.com/doc/current/service_container/shared.html

**Reply by ro0NL** [reply]
ofc. you don't need to use the main app's container. You could abstract it in your own service layer by constructing a dedicated container(builder) (not sure about perf tho :))

**Reply by exteon** [reply]
If I get what you are saying right, then as far as Symfony is concerned, Dom1... Dom200 will be services?

If so, yeah, I'm worried about the performance impact? I mean, all I really need is: 

Parse constructor def -> Resolve (trivial, a class-to-instance map) -> Cache -> Send to instance constructor

I mean, I can have all the necessary services already instanced and possibly adapted/layered in the container... I just need the injection part, and the whole service concept seems to involve a whole lot of overhead on top of that.

**Reply by exteon** [reply]
> while im generally skeptical about (domain) models, depending on services (usually it's in reverse; services provide/take models)

Yeap, my point exactly, that's why i need an abstract model builder in the service that can do some DI, so I can eliminate the reference from model to service /container

**Reply by ro0NL** [reply]
> Parse constructor def -> Resolve (trivial, a class-to-instance map) -> Cache -> Send to instance constructor

that's what the DependencyInjection component does yes :) where constructor parsing is done by "autowiring".

in the end, a service container relies on a object factory. otherwise, no. There's no standard object factory 

as mentioned, you could leverage the/a service container as general purpose factory.

**Reply by ahundiak** [reply]
You probably should not be overly worried about performance.  Over the years the cached Symfony container has undergone many optimizations.  It's probably faster than something you might write by hand.  Consider creating one of your services then looking at the generated code under var/cache.  At first blush it may seem a complete mess but it's really just a big array that points to individual factories for each service.

And as we discussed a bit on stackoverflow, you can refine things slightly by grouping your domain objects with a service locator.  Only your objects will be accessible via the locator so you are no longer dealing directly with the global container.

But start by convincing yourself that Symfony container itself will meet your needs as stated.

**Reply by ro0NL** [reply]
> as far as Symfony is concerned, Dom1... Dom200 will be services?

that's the issue, idealy we dont want to leak such factorized models in our main app's container.

i'd even try if `new PsrContainerBasedFactoryImplem(new ContainerBuilder())` without compilation is considerable.

**Reply by exteon** [reply]
I'm taking a break from Symfony to look at PHP-DI at the moment... the idea is I will be pretty much using Symfony for everything else, so it would be a danggood thing that I do not add yet-another-code-dependency... But alas, gotta read the KISS principle as "keep-it-simple", only then "keep-it-short". 90% of the time, simple === short, but not always

**Comment by javiereguiluz** [comment]
@exteon please don't forget to "Mark as answer" the comment that replies to your question more accurately. Thanks!

**Reply by exteon** [reply]
Thanks for the reminder and sorry for the delay, but as always, life is what happens when you had other plans... Alas this problem is deferred to an undefined future. Honestly no answer in the Symfony realm is good enough, as using the service container as a model factory would mean relying on a very frail contract, introducing the worst kind of technical debt, which is using things in ways they are not meant to. So the question just transformed to "what solution outside symfony would work best", which frankly is beyond the scope of the question itself. So I guess we can close this with an "unanswered" answer 🙂
